lib/Transform/Clang/LoopSwapping.cpp

```diff
@@ -0,0 +1,546 @@
+  std::vector<std::vector<SourceRange>> mRangePairs;
+  std::vector<std::vector<Loop *>> mLoopPairs;
+  std::stack<ForStmt *> *mForStack;
+};
+
+class ClangLoopSwappingInfo final : public PassGroupInfo {
+  void addBeforePass(legacy::PassManager &Passes) const override {
+    addImmutableAliasAnalysis(Passes);
+    addInitialTransformations(Passes);
+    Passes.add(createAnalysisSocketImmutableStorage());
+    Passes.add(createDIMemoryTraitPoolStorage());
+    Passes.add(createDIMemoryEnvironmentStorage());
+    Passes.add(createDIEstimateMemoryPass());
+    Passes.add(new ClangLoopSwappingServer);
+    Passes.add(createAnalysisWaitServerPass());
+  }
+
+  void addAfterPass(legacy::PassManager &Passes) const override {
+    Passes.add(createAnalysisReleaseServerPass());
+    Passes.add(createAnalysisCloseConnectionPass());
+  }
+};
+
+}
+
+void ClangLoopSwapping::initializeProviderOnServer() {
+  ClangLoopSwappingServerProvider::initialize<GlobalOptionsImmutableWrapper>(
+      [this](GlobalOptionsImmutableWrapper &Wrapper) {
+        Wrapper.setOptions(mGlobalOpts);
+      });
+  auto R = mSocket->getAnalysis<
+      DIMemoryEnvironmentWrapper, DIMemoryTraitPoolWrapper>();
+  assert(R && "Immutable passes must be available on server!");
+  auto *DIMEnvServer = R->value<DIMemoryEnvironmentWrapper *>();
+  ClangLoopSwappingServerProvider::initialize<DIMemoryEnvironmentWrapper>(
+      [DIMEnvServer](DIMemoryEnvironmentWrapper &Wrapper) {
+        Wrapper.set(**DIMEnvServer);
+      });
+  auto *DIMTraitPoolServer = R->value<DIMemoryTraitPoolWrapper *>();
+  ClangLoopSwappingServerProvider::initialize<DIMemoryTraitPoolWrapper>(
+      [DIMTraitPoolServer](DIMemoryTraitPoolWrapper &Wrapper) {
+        Wrapper.set(**DIMTraitPoolServer);
+      });
+      
+}
+
+std::vector<const DIAliasTrait *> ClangLoopSwapping::GetLoopTraits(MDNode *LoopID) {
+  auto DepItr = DIDepInfo->find(LoopID);
+  assert(DepItr != DIDepInfo->end() && "Loop must be analyzed!");
+  auto &DIDepSet = DepItr->get<DIDependenceSet>();
+  DenseSet<const DIAliasNode *> Coverage;
+  accessCoverage<bcl::SimpleInserter>(DIDepSet, *DIAT, Coverage,
+                                      mGlobalOpts->IgnoreRedundantMemory);
+  std::vector<const DIAliasTrait *> traits;
+  for (auto &TS : DIDepSet) {
+    auto node = TS.getNode();
+    if (!Coverage.count(node))
+      continue;
+    traits.push_back(&TS);
+  }
+  return traits;
+}
+
+bool ClangLoopSwapping::HasSameReductionKind(std::vector<const DIAliasTrait *> &traits0,
+                        std::vector<const DIAliasTrait *> &traits1,
+                        SpanningTreeRelation<DIAliasTree *> &STR) {
+  for (auto TS0: traits0) {
+    auto *node0 = const_cast<DIAliasNode *>(TS0->getNode());
+    MemoryDescriptor Dptr0 = *TS0;
+    for (auto TS1: traits1) {
+      auto *node1 = const_cast<DIAliasNode *>(TS1->getNode());
+      MemoryDescriptor Dptr1 = *TS1;
+      if (STR.isEqual(node0, node1) && Dptr0.is<trait::Reduction>() &&
```

Comment:
Для проверки на равенство достаточно использовать `==`, можно не строить остовное дерево (STR).
```suggestion
      if (node0 == node1 && Dptr0.is<trait::Reduction>() &&
```